enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1012 @@
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+#### Removing a deprecated feature
+
+Does not apply.
+
+### Upgrade / Downgrade Strategy
+
+Does not apply.
+
+### Version Skew Strategy
+
+Does not apply.
+
+### Operational Aspects of API Extensions
+
+All the API reads will happen via informers, thus the API impact will be minimal.
+
+#### Failure Modes
+
+The feature is compartmentalized in order to protect the default cluster
+network. It will not impact it in any way.
+
+Performance impacts will impact IP address assignment in the network's scope,
+thus other networks are also not impacted.
+
+#### Support Procedures
+
+If the IPAM plugin fails to generate an `IPAMClaim` for the VM, an event will
+be thrown, and the error logged.
+
+We can also create some `condition`s to highlight some specific errors in the
+`IPAMClaim` - e.g. IP pool is full, etc.
+
+## Implementation History
+
+TODO
+
+## Alternatives
+
+### Have OVN-Kubernetes create the `IPAMClaim`
+This alternative does **not** require the `IPAMClaim` CRD to have a status
+sub-resource.
+
+The main difference is KubeVirt instructs the CNI plugin via CNI-Args (over the
+network-selection-elements) the pod being created is requesting an IP on behalf
+of another entity - i.e. the VM. The network selection element would look like:
+```json
+{
+  "name": "tenantred",
+  "namespace": "default",
+  "interface": "pod16367aacb67",
+  "cni-args": {
+    "ipamclaim.cni.cncf.io/ownerID": "a0790345-4e84-4257-837a-e3d762d191ab",
+    "ipamclaim.cni.cncf.io/ownerName": "vm-a",
+    "ipamclaim.cni.cncf.io/ownerType": "VirtualMachine",
+    "ipamclaim.cni.cncf.io/ownerVersion": "kubevirt.io/v1"
+  }
+}
+```
+
+When the CNI plugin notices this information in the network selection element,
+it allocates an IP address from the pool, and then creates the CR with the IPs
+directly in the spec - keep in mind there's no reason for this alternative to
+feature the status sub-resource.
+
+The CNI plugin will also create the `IPAMClaim` with the owner reference
+information, which was passed via CNI-Args. The CR would look like:
+```yaml
+apiVersion: "ipamclaims.k8s.cni.cncf.io/v1alpha1"
+kind: IPAMClaim
+metadata:
+  name: vm-a.tenantred-attachment
+  namespace: ns1
+  ownerReferences:
+  - apiVersion: kubevirt.io/v1
+    kind: VirtualMachine
+    name: vm-a
+    uid: a0790345-4e84-4257-837a-e3d762d191ab
+spec:
+  network: tenantred
+  interface: pod16367aacb67
+  ips:
+  - 10.128.20.8/24
+  - fd10:128:20::8/64
+```
```

Comment:
I think that the main advantage of this approach would be that it would not require an IPAMClaim at all, meaning, it could be a CNI implementation detail how to handle this, it could do so with a CRD of its own and that is something that I probably like better because it does give more flexibility to the CNI itself.

Unless kubevirt needs the actual IPAMClaim for something but I have not seen that in this enhancement.

Reply:
Could you add some (rough) detail into how would that tie to the CNI's IP pool ? I just can't picture it.

Reply:
So the CNI would get something like
```
    "ipamclaim.cni.cncf.io/ownerID": "a0790345-4e84-4257-837a-e3d762d191ab",
    "ipamclaim.cni.cncf.io/ownerName": "vm-a",
    "ipamclaim.cni.cncf.io/ownerType": "KubevirtIPAMClaim",
    "ipamclaim.cni.cncf.io/ownerVersion": "kubevirt.io/v1"
```
kubevirt would be completely in charge of managing the lifecycle of this `KubevirtIPAMClaim` according to its needs. The CNI would just couple the IP assignment to the lifecycle of this `KubevirtIPAMClaim`. It could do so having its own CRD which would have an owner reference set to this `KubevirtIPAMClaim`.

So it would work just mostly the same but instead of having a single central shared CRD, each component would have its own CRD which could potentially be more flexible.

I just wanted to elaborate this detail further because I think its the main advantage of this proposal. I guess it has its disadvantages as well but I think it would meet all kubevirt requirements anyway.

Reply:
I see. I understand the advantage (allowing each CNI to be more flexible) but I just don't see use cases for it and there is no current ask for it. 

I would prefer to stick to the simpler (less pieces) implementation unless there's a clear motive (use cases) requiring this extra customization. 

But still, as you say: it's mostly the same. Do you think this deserves a couple added paragraphs ?